### PR TITLE
Migrate fetching archived flow cells from NAS-9 to Hasta

### PR DESCRIPTION
**Pull request by barrystokman**
## Description
Fetching archived flow cells is handled by bash scripts running on NAS-9. We don't 'own' NAS-9,  so we should have no fucntionality on there at all. This PR migrates the fetch flow cell functionality to python on Hasta.

### Added
Fetching of archived flow cell/key pairs and handling subsequent key decryption, flow cell decryption, flow cell archive extraction, preparing the retrieved flow cell for automated demux start and clean up of files used in the retrieval process

### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [x] TC1: fetch specified flow cell in dry run mode `cg backup fetch-flow-cell -f H2MFNDRX2 -d`
- [x] TC2: fetch unspecified flow cell in dry run mode, no flow cell requested `cg backup fetch-flow-cell  -d`
- [x] TC3: fetch unspecified flow cell in dry run mode, flow cell requested `cg backup fetch-flow-cell -d`
- [x] TC4: fetch specified flow cell `cg backup fetch-flow-cell -f H2MFNDRX2`
- [x] TC5: fetch unspecified flow cell, no flow cell requested `cg backup fetch-flow-cell`
- [x] TC6: fetch unspecified flow cell, flow cell requested `cg backup fetch-flow-cell`

### Test preconditions:
- [x] TC1: set the status of flow cell  `H2MFNDRX2` to `removed`, remove any files related to `H2MFNDRX2` from `/home/proj/stage/flowcells/novaseq/runs/`
- [x] TC2: no flow cell should have the status `requested`
- [x] TC3: set the status of flow cell `H2MFNDRX2` to `requested`
- [x] TC4: set the status of flow cell `H2MFNDRX2` to `removed`
- [x] TC5: N/A
- [x] TC6: remove any existing run directory, set the status of flow cell `H2MFNDRX2` to `requested`

### Expected test outcome:
- [x] TC1: dry run should display flow cell and key found on pdc, not retrieval should start, not flow cell status should be set
- [x] TC2: dry run should display `no flow cell` and exit
- [x] TC3: see TC1
- [x] TC4: the key and flow cell should be retrieved from PDC to the flow cell run directory `/home/proj/stage/flowcells/novaseq/runs/`. The key should be drecrypted, the flow cell should be decrypted, the flow cell should be extracted into the flow cell run directory `220325_A00621_0651_BH2MFNDRX2`, the file `RTAComplete.txt` should be created in the run directory, the retrieved, decrypted and extracted files should be removed.
- [x] TC5: dry run should display `no flow cell` and exit
- [x] TC6: see TC4

## Review
- [x] code approved by PG
- [x] tests executed by BS
- [x] "Merge and deploy" approved by KN, PG
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan
- [x] deploy with https://github.com/Clinical-Genomics/servers/pull/813
- [ ] when deployed, NAS-9 can be decomissioned 🥳

### CODE REVIEW

**Comment by barrystokman** [comment]
TC1: ✔️ 
image
TC2: ✔️ 
image
TC3: ✔️ 
image

**Comment by barrystokman** [comment]
Please note that when retrieving files fomr PDC to Hasta SELinux will raise a warning (status code 8) when attempting to write file attributes. This causes an exception in `cg.utils.Process`. If the return code of the called process is 8, I'm logging a warning, but the process continues, as intended. This error has no consequences for this process.

File: cg/meta/backup/backup.py [hunk 1]
```diff
@@ -62,33 +77,211 @@ def fetch_flow_cell(
...
+                    self.status.commit()
+                raise error
 
-        return get_elapsed_time(start_time=start_time)
+    def query_pdc_for_flow_cell(self, flow_cell_id) -> list:
+        """Query PDC for a given flow cell id"""
+        search_pattern = EncryptionDirsAndFiles.ENCRYPT_DIR + ASTERISK + flow_cell_id + ASTERISK
+        self.pdc.query_pdc(search_pattern=search_pattern)
+        query: list = self.pdc.process.stdout.split(NEW_LINE)
+        return query
```
Comment (karlnyr) [inline cg/meta/backup/backup.py:248]: This query could be ordered by date and return only the latest entries?

File: tests/meta/backup/conftest.py [hunk 2]
```diff
@@ -13,3 +13,42 @@ def fixture_spring_file_path() -> Path:
...
+        "Accessing as node: SLLCLINICAL",
+        "             Size  Archive Date - Time    File - Expires on - Description",
+        "             ----  -------------------    -------------------------------",
+        "           607  B  03/28/2022 11:59:54    "
+        "/path/to/archived/encryption_key.key.gpg Never "
+        "Archive Date: 03/28/2022",
+        "43,868,543,141  B  03/28/2022 11:53:04    "
+        "/path/to/archived/flow_cell.tar.gz.gpg Never "
+        "Archive Date: 03/28/2022",
+    ]
```
Comment (karlnyr) [inline tests/meta/backup/conftest.py:54]: To this we could add another copied entry with an earlier date for ToF for multiple entries of an archive

**Review by karlnyr — COMMENTED** [review]
It looks like this is good to me, the one case we would have to consider is if there are somehow multiple entries. And of course, testing 👍

**Review by patrikgrenfeldt — APPROVED** [review]
I suggest we merge as is to avoid having to retest because we change the logic somehow since the main author left and then make suggested improvements to the new codebase

File: cg/constants/encryption.py [hunk 3]
```diff
@@ -56,3 +56,9 @@ class GPGParameters(ListEnum):
     OUTPUT_PARAMETER: list = [
         "-o",
     ]
+
+
+class EncryptionDirsAndFiles(StrEnum):
+    """Encryption related dirs and files"""
+
+    ENCRYPT_DIR: str = "/home/hiseq.clinical/ENCRYPT/"
```
Comment (Mropat) [inline cg/constants/encryption.py:64]: this shouldnt be hardcoded in cg, but come from config file in servers or envvars

File: cg/meta/backup/backup.py [hunk 4]
```diff
@@ -62,33 +77,211 @@ def fetch_flow_cell(
...
 
         flow_cell_obj.status = FlowCellStatus.PROCESSING
-        if not dry_run:
+        if not self.dry_run:
             self.status.commit()
             LOG.info("%s: retrieving from PDC", flow_cell_obj.name)
 
-        start_time = get_start_time()
+        try:
+            pdc_flow_cell_query: list = self.query_pdc_for_flow_cell(flow_cell_obj.name)
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/meta/backup/backup.py:85]: Created issue with this comment: https://github.com/Clinical-Genomics/cg/issues/1515

File: cg/meta/backup/backup.py [hunk 5]
```diff
@@ -7,25 +7,42 @@
...
     """Class for retrieving FCs from backup"""
 
-    def __init__(self, status: Store, pdc_api: PdcAPI, root_dir: dict):
+    def __init__(
+        self,
+        encryption_api: EncryptionAPI,
+        status: Store,
+        tar_api: TarAPI,
+        pdc_api: PdcAPI,
+        root_dir: dict,
```
[the remark this answers is not in the document]
Comment (patrikgrenfeldt) [reply cg/meta/backup/backup.py:36]: Created issue with this comment: https://github.com/Clinical-Genomics/cg/issues/1515

File: cg/meta/backup/backup.py (hunk 1 above)
Comment (patrikgrenfeldt) [reply cg/meta/backup/backup.py:248]: Created issue with this comment: https://github.com/Clinical-Genomics/cg/issues/1515

File: tests/meta/backup/conftest.py (hunk 2 above)
Comment (patrikgrenfeldt) [reply tests/meta/backup/conftest.py:54]: Created issue with this comment: https://github.com/Clinical-Genomics/cg/issues/1515

File: cg/constants/encryption.py (hunk 3 above)
Comment (patrikgrenfeldt) [reply cg/constants/encryption.py:64]: Created issue with this comment: https://github.com/Clinical-Genomics/cg/issues/1515

**Comment by patrikgrenfeldt** [comment]
Deployed on Hasta:

**Comment by patrikgrenfeldt** [comment]
> @patrikgrenfeldt does this PR require anything to be changed in servers, or is this using the old archiving scripts and cronjobs?

We merged a servers PR earlier for tar

**Comment by Mropat** [comment]
@patrikgrenfeldt what about **archiving scripts and cronjobs**?

**Comment by patrikgrenfeldt** [comment]
As I understand it; with this PR we are stopping sending "fetch-jobs" to the NAS and instead execute the job directly on the HPC. The scheduling running on the HPC has not been changed and should not have to be changed for this to work.

**Pull request merged**